### PR TITLE
Bring discovery on par with the one in docker/swarm

### PR DESCRIPTION
**Pull request by vieux**
A couple of fixes/patches were made in `docker/swarm/discovery` by @abronan 

Since out plan is to use the discovery from engine in swarm to avoid
duplicate code, I'm porting the changes here.

Related to https://github.com/docker/swarm/issues/1594

### CODE REVIEW

**Comment by abronan** [comment]
SGTM as it's just bringing the two on par with one another.

To add more context to those fixes:

- The first one allows to customize the discovery path for swarm to match the one of Engine (and thus avoid two discovery systems operating simultaneously if Engine is used for the discovery instead of `swarm join`).

- The second one is necessary if the key does not exist yet in the store. It just adds a clause to the **Key** creation using *libkv* to make sure that it is created as a special **Directory** instead of a regular Key (for etcd).

**Comment by calavera** [comment]
LGTM

**Comment by vieux** [comment]
@abronan thanks for giving a better explanation ( I kept your commit messages)

**Comment by vieux** [comment]
tests: 7 / 7 :green_heart:

**Comment by icecrime** [comment]
LGTM :+1: Moving to docs review (I'm assuming the configurable path needs some doc?).

**Comment by vdemeester** [comment]
ping @thaJeztah but I don't think this needs docs-review though :stuck_out_tongue_closed_eyes:

**Comment by vieux** [comment]
I just added a small doc change @thaJeztah @moxiegirl

File: docs/reference/commandline/daemon.md [hunk 1]
```diff
@@ -603,6 +603,10 @@ The currently supported cluster store options are:
     private key is used as the client key for communication with the
     Key/Value store.
 
+*  `kv.path`
+
+    Specifies the path you want to use in the Key/Value store. (default: docker/nodes)
```
Comment (thaJeztah) [inline docs/reference/commandline/daemon.md]: Specifies the path in the Key/Value store. If not configured, the default value is 'docker/nodes'.

**Comment by thaJeztah** [comment]
ping @vieux; left a suggestion, can you also squash your commits?

**Comment by vieux** [comment]
@thaJeztah done

**Comment by thaJeztah** [comment]
docs LGTM, thanks!

Failing docs is not related;

    FATAL: 2016/01/08 mkdir /docs/public/engine/extend/images/: no such file or directory
ERRORS = 1

**Comment by thaJeztah** [comment]
hm. I'll restart Janky

```
22:18:42 time="2016-01-08T22:18:30Z" level=error msg="Download failed, retrying: simulating retry" 
22:18:42 --- FAIL: TestSuccessfulDownload (5.23s)
22:18:42 	download_test.go:300: missing or wrong progress output for id3 (got: 9)
22:18:42 time="2016-01-08T22:18:36Z" level=error msg="Upload failed, retrying: simulating retry" 
22:18:42 FAIL
```

**Comment by vieux** [comment]
@thaJeztah can we merge ?

**Comment by thaJeztah** [comment]
@vieux; I restarted the experimental build, but if you think it's safe to go, please do :+1:

**Comment by vieux** [comment]
@thaJeztah I'm not maintainer anymore :cry:

**Comment by thaJeztah** [comment]
@vieux I trust you :smile: 

otherwise it's just waiting for the experimental test to finish, then we'll merge :)

**Pull request merged**